### PR TITLE
Support logits to frozen Bernoulli distribution

### PR DESCRIPTION
**Pull request by fehiepsi**
This PR supports a kwarg `is_logits` to Bernoulli distribution. By default, is_logits=False and it will be so for non-frozen distribution. For frozen distribution, by adding `is_logits` argument, the arg `p` will be considered as a logit, instead of prob.

@neerajprad Pls find my reasoning in comments. There might be some edge cases which I have not figured it out yet, so please let me know your thought about this change. :)

### CODE REVIEW

File: test/test_distributions.py [hunk 1]
```diff
@@ -140,7 +141,7 @@ def test_logprob(jax_dist, loc_scale):
 
 @pytest.mark.parametrize('jax_dist, dist_args', [
     (dist.bernoulli, (0.1,)),
-    (dist.bernoulli, (np.array([0.3, 0.5]))),
+    (dist.bernoulli, (np.array([0.3, 0.5]),)),
```
Comment (fehiepsi) [inline test/test_distributions.py]: this is typo

File: numpyro/distributions/util.py [hunk 2]
```diff
@@ -273,6 +273,12 @@ def entr(p):
     return e
 
 
+def binary_cross_entropy_with_logits(x, y):
+    # compute -y * log(sigmoid(x)) - (1 - y) * log(1 - sigmoid(x))
+    # Ref: https://www.tensorflow.org/api_docs/python/tf/nn/sigmoid_cross_entropy_with_logits
+    return np.clip(x, 0) - x * y + np.log1p(np.exp(-np.abs(x)))
```
Comment (fehiepsi) [inline numpyro/distributions/util.py]: nit: reorder the terms such as terms involving only `x` is grouped (to save a bit of time when y has batch dimensions).

File: numpyro/distributions/distribution.py [hunk 3]
```diff
@@ -67,11 +67,11 @@ def rvs(self, *args, **kwargs):
         # assert that rng is PRNGKey and not mtrand.RandomState object from numpy.
         assert _is_prng_key(rng)
         kwargs['random_state'] = onp.random.RandomState(rng)
-        sample = super(osp_stats.rv_discrete, self).rvs(*args, **kwargs)
+        sample = super(jax_discrete, self).rvs(*args, **kwargs)
```
Comment (fehiepsi) [inline numpyro/distributions/distribution.py]: I think this is a typo too

File: numpyro/distributions/discrete.py [hunk 4]
```diff
@@ -33,8 +34,39 @@ def _entropy(self, n, p):
 
 
 class _bernoulli_gen(jax_discrete, bernoulli_gen):
+    def __new__(cls, *args, **kwargs):
+        return super(_bernoulli_gen, cls).__new__(cls)
```
Comment (fehiepsi) [inline numpyro/distributions/discrete.py]: [rv_discrete](https://github.com/scipy/scipy/blob/master/scipy/stats/_distn_infrastructure.py#L2690) has a signature which prevents us from adding additional keyword `is_logits`.

File: numpyro/distributions/discrete.py [hunk 5]
```diff
@@ -33,8 +34,39 @@ def _entropy(self, n, p):
...
+
+    def __init__(self, *args, **kwargs):
+        self.is_logits = kwargs.pop("is_logits", False)
+        super(_bernoulli_gen, self).__init__(*args, **kwargs)
+
+    def freeze(self, *args, **kwargs):
+        self._ctor_param.update(is_logits=kwargs.pop("is_logits", False))
```
Comment (fehiepsi) [inline numpyro/distributions/discrete.py]: We update `_ctor_param` so that [frozen distribution](https://github.com/scipy/scipy/blob/master/scipy/stats/_distn_infrastructure.py#L427) makes correct signatures for its self.dist.

File: test/test_mcmc.py [hunk 6]
```diff
@@ -44,10 +44,8 @@ def test_logistic_regression():
         def potential_fn(beta):
             coefs_mean = np.zeros(dim)
             coefs_lpdf = dist.norm(coefs_mean, np.ones(dim)).logpdf(beta)
-            probs = np.clip(expit(np.sum(beta * data, axis=-1)),
-                            a_min=np.finfo(np.float32).tiny,
-                            a_max=(1 - np.finfo(np.float32).eps))
-            y_lpdf = dist.bernoulli(probs).logpmf(labels)
+            logits = np.sum(beta * data, axis=-1)
+            y_lpdf = dist.bernoulli(logits, is_logits=True).logpmf(labels)
```
Comment (neerajprad) [inline test/test_mcmc.py]: Nice!

File: numpyro/distributions/discrete.py (hunk 5 above)
Comment (neerajprad) [reply numpyro/distributions/discrete.py]: This is a nice work-around.

File: test/test_distributions.py [hunk 7]
```diff
@@ -192,3 +193,20 @@ def test_discrete_logpmf_args_check(jax_dist, dist_args):
...
         jax_dist.logpmf(sample_oos, *dist_args)
+
+
+@pytest.mark.parametrize('jax_dist, dist_args', [
+    (dist.bernoulli, (0.1,)),
+    (dist.bernoulli, (np.array([0.3, 0.5]),)),
+], ids=idfn)
+def test_discrete_with_logits(jax_dist, dist_args):
+    rng = random.PRNGKey(0)
+    logit_args = dist_args[:-1] + (logit(dist_args[-1]),)
```
Comment (neerajprad) [inline test/test_distributions.py]: I think you probably just meant `logit(dist_args)`?

Comment (fehiepsi) [reply test/test_distributions.py]: Yes, I mean so. I expect more distributions will need this as mentioned in https://github.com/pyro-ppl/numpyro/issues/51, so this just serves for future tests.

**Comment by fehiepsi** [comment]
> we should probably only support frozen distributions and keep our code simple

Indeed, I like the current state because we inherit many mechanisms from scipy. Adding more distribution won't be much extra work (except multivariate ones, where each dist has its own mechanism) so it is also convenience for other developers. We indeed do not interact with frozen dists because all the logics belong to unfrozen dists. Except for pyro models (which requires frozen dist), users can define a custom objective with their familiar coding style. But I do agree that we only support extra features for frozen distributions. One of the extra features which I have in mind now is having `scale_tril` arg for multivariate normal distribution.

File: test/test_distributions.py (hunk 7 above)
Comment (neerajprad) [reply test/test_distributions.py]: @fehiepsi - I just meant why do we need to do `dist_args[:-1] + (logit(dist_args[-1]),)` instead of just using `(logit(dist_args),)`?

Comment (fehiepsi) [reply test/test_distributions.py]: For example, in binom/multinomial distributions, we have 2 args: n and p. So we only want to apply `logit` to `p`.
